jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java

```diff
@@ -64,48 +65,57 @@ private enum State
         /**
          * This callback is just about to call {@link #process()},
          * or within it, or just exited from it, either normally
-         * or by throwing.
+         * or by throwing. Further actions are waiting for the
+         * {@link #process()} method to return.
          */
         PROCESSING,
 
         /**
-         * Method {@link #process()} returned {@link Action#SCHEDULED}
-         * and this callback is waiting for the asynchronous sub-task
-         * to complete.
+         * The asynchronous sub-task was completed either with
+         * a call to {@link #succeeded()} or {@link #failed(Throwable)}, whilst in
+         * {@link #PROCESSING} state.  Further actions are waiting for the
+         * {@link #process()} method to return.
          */
-        PENDING,
+        PROCESSING_CALLED,
 
         /**
-         * The asynchronous sub-task was completed successfully
-         * via a call to {@link #succeeded()} while in
-         * {@link #PROCESSING} state.
+         * The was {@link #abort(Throwable) aborted} whilst in
+         * {@link #PROCESSING} state. Further actions are waiting for the
+         * {@link #process()} method to return.
          */
-        CALLED,
+        PROCESSING_ABORTED,
```

Comment:
I think adding states is going to make it too complicated.

We already have a boolean for re-iteration, that we check only in relevant places.

We should use a similar approach, checking `_aborted` in relevant places.

We gave it a quick go, and the code turned out to be much more readable and understandable, because we leave the current state machine as is, and only handle `_aborted` in these cases:

* in succeeded() and failed() when state is PENDING (because we need to "resume" and call onComplete(Throwable)
* after we return from process()

This keep the guarantee that there is only one thread calling callbacks.

And we must use the same trick used in `Callback.Abstract` to make sure onAbort() is notified before onCompleted(Throwable).
Which could be done with a special `AbortException` that holds the failure and whether the onAbort() has been called already.

Reply:
This is the approach I used in `Callback.Abstract`, just turned out it didn't need to be an exception holding and exception as success is static. 
I'll look at doing same here.       Do you push your code anywhere?

Reply:
Dang! it will make my `testAbortInEveryState` less useful, as I currently exhaustively test abort there.   I'm not opposed to removing these states if it does simplify the code... but do we think there is actually anything wrong with this current implementation?

Reply:
No we wrote the code on-the-fly to verify how complicated it would be, but we did not push it.

Reply:
@sbordet @lorban well thought experiments are always easier than reality!!!!     It proved to look less complex, but to actually be a lot more complex and difficult to test.    I think it is done and it does look simpler, but I think it is a lot more fragile as the subtle issues are now hidden behind complex `if(){}elseif(){}else{}` logic.  

I've written a fairly thorough unit test, but it was not easy.   Things like close are a bit strange... I've modelled that as an abort that cannot be reset.